job-runner/server/server.go

```diff
@@ -61,47 +61,59 @@ func (s *Server) API() *api.API {
 
 // --------------------------------------------------------------------------
 
-func loadConfig(appCtx *app.Context) error {
+// Catch TERM and INT signals to gracefully shut down the Job Runner
+func (s *Server) waitForShutdown() {
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	<-sigChan
+
+	// API + traversers watch shutdownChan
+	close(s.shutdownChan)
```

Comment:
This will send a signal to the API and traversers to begin the shutdown process, but does it wait for them to finish shutting down?

Reply:
Hmm, you're right, this doesn't really wait for anything to shut down - just catches the SIGINT/SIGTERM, which stops the JR from getting shut down until it eventually gets killed (by which point, ideally, the traversers have had a chance to shut down). It's better, though, to have the JR shut down on its own once the traversers + API are all done shutting down - I'll add code for that.